text/0001-bee-network/0001-bee-network.md

```diff
@@ -159,37 +150,46 @@ impl Router {
 
         let id = self.gen_peer_id();
 
-        let conn = match config {
-            ProtocolConfig::Tcp(config) => TcpProtocol::setup(protocol::ProtocolConfig::Tcp(config))?,
-        };
-
-        self.connections.insert(id, protocol::ProtocolType::Tcp(conn));
+        match config {
+            ProtocolConfig::Tcp(x) => {
+                let protocol = TcpProtocol::setup(protocol::ProtocolConfig::Tcp(x))?;
+                self.connections.insert(id, Box::new(protocol));
+            }
+            ProtocolConfig::Udp(x) => {
+                let protocol = UdpProtocol::setup(protocol::ProtocolConfig::Udp(x))?;
+                self.connections.insert(id, Box::new(protocol));
+            }
+        }
 
         Ok(id.clone())
 
     }
 
     pub fn send_to(&self, id: PeerId, msg: &[u8]) -> Result<(), Error> {
 
-        match self.connections.get_mut(&id).ok_or(ErrorKind::NotFound)?.deref_mut() {
-            ProtocolType::Tcp(tcp) => Ok(tcp.send(msg)?),
-        }
+        let mut map_ref_mut= self.connections.get_mut(&id).ok_or(ErrorKind::NotFound)?;
+        let mut box_ref_mut = map_ref_mut.deref_mut();
+        let protocol_ref_mut = box_ref_mut.deref_mut();
+
+        Ok((protocol_ref_mut.send(msg)?))
 
     }
 
     pub fn recv(&self, id: PeerId) -> Result<Vec<u8>, Error> {
 
-        match self.connections.get_mut(&id).ok_or(ErrorKind::NotFound)?.deref_mut() {
-            ProtocolType::Tcp(tcp) => Ok(tcp.recv()?),
-        }
+        let mut map_ref_mut= self.connections.get_mut(&id).ok_or(ErrorKind::NotFound)?;
```

Comment:
You should be able to just use `*&*&foo` here

Reply:
Do you mean `&*&self.connections.get_mut(&id).ok_or(ErrorKind::NotFound)?`? It returns `&DashMapRefMut<PeerId,Box<Protocol>>`